src/containerapp/azext_containerapp/_utils.py

```diff
@@ -1714,3 +1726,161 @@ def format_location(location=None):
     return location
+
+
+def is_docker_running():
+    try:
+        # Run a simple 'docker stats --no-stream' command to check if the Docker daemon is running
+        command = ["docker", "stats", "--no-stream"]
+        process = subprocess.Popen(command, stdout=subprocess.PIPE, stderr=subprocess.PIPE)
+        _, _ = process.communicate()
+        return process.returncode == 0
+    except Exception:
+        return False
+
+
+def get_pack_exec_path():
+    try:
+        dir_path = os.path.dirname(os.path.realpath(__file__))
+        bin_folder = os.path.join(dir_path, "bin")
+        if not os.path.exists(bin_folder):
+            os.makedirs(bin_folder)
+
+        pack_cli_version = "v0.29.0"
+        exec_name = "pack"
+        compressed_download_file_name = f"pack-{pack_cli_version}"
+        host_os = platform.system()
+        if host_os == "Windows":
+            compressed_download_file_name = f"{compressed_download_file_name}-windows.zip"
+            exec_name = "pack.exe"
+        elif host_os == "Linux":
+            compressed_download_file_name = f"{compressed_download_file_name}-linux.tgz"
+        elif host_os == "Darwin":
+            compressed_download_file_name = f"{compressed_download_file_name}-macos.tgz"
+        else:
+            raise Exception(f"Unsupported host OS: {host_os}")
+
+        exec_path = os.path.join(bin_folder, exec_name)
+        if os.path.exists(exec_path):
+            return exec_path
+
+        # Attempt to install the pack CLI
+        url = f"https://github.com/buildpacks/pack/releases/download/{pack_cli_version}/{compressed_download_file_name}"
+        with urlopen(url) as req:
+            compressed_file = io.BytesIO(req.read())
+            if host_os == "Windows":
+                with zipfile.ZipFile(compressed_file) as zip_file:
+                    for file in zip_file.namelist():
+                        if file.endswith(exec_name):
+                            with open(exec_path, "wb") as f:
+                                f.write(zip_file.read(file))
+            else:
+                with tarfile.open(fileobj=compressed_file, mode="r:gz") as tar:
+                    for tar_info in tar:
+                        if tar_info.isfile() and tar_info.name.endswith(exec_name):
+                            with open(exec_path, "wb") as f:
+                                f.write(tar.extractfile(tar_info).read())
+
+        # Add executable permissions for the current user if they don't exist
+        if not os.access(exec_path, os.X_OK):
+            st = os.stat(exec_path)
+            os.chmod(exec_path, st.st_mode | stat.S_IXUSR)
+
+        return exec_path
+    except Exception as e:
+        # Swallow any exceptions thrown when attempting to install pack CLI
+        logger.warning(f"Failed to install pack CLI: {e}\n")
+
+    return None
+
+
+def patchable_check(repo_tag_split: str, oryx_builder_run_img_tags, inspect_result):
```

Comment:
Since there are some join and slice operations, it is suggested to add some comments to explain this function.

Reply:
Updated: added comments to explain the general workflow and string arithmetic.